File: leetcodes/322. Coin Change/coinChange.js
/**
 * @param {number[]} coins
 * @param {number} amount
 * @return {number}
 */
//can't use DP = new Array(coins.length) since it always counts the max of the largest coins so it is easy to get -1
var coinChange = function(coins, amount) {
  if (amount === 0) return 0
  coins.sort((a,b)=> a-b)
  //time O(nlogn)
  let n = coins.length;
  let DP = new Array(amount+1).fill(Infinity)
  DP[0] = 0
  //space O(amount)
  for (let i = 0; i <= amount ; i++) {
    coins.forEach(coin => {
      if (coin <= i) {
        DP[i] = Math.min(DP[i], DP[i - coin] + 1)
      }
    });
  }
  //time O(amount * n)
  return DP[amount] === Infinity ? -1 : DP[amount]
  //bottom up from 0 to amount
};

File: leetcodes/322. Coin Change/coinChange.java
public class Solution {
  // O(amount) space, O(n*amount) time complexity
  public int coinChange(int[] coins, int amount) {
    if (coins == null || coins.length == 0)
      return -1;

    if (amount <= 0)
      return 0;

    int dp[] = new int[amount + 1];
    for (int i = 1; i < dp.length; i++) {
      dp[i] = Integer.MAX_VALUE;
    }

    for (int am = 1; am < dp.length; am++) {
      for (int i = 0; i < coins.length; i++) {
        if (coins[i] <= am) {
          int sub = dp[am - coins[i]];
          if (sub != Integer.MAX_VALUE)
            dp[am] = Math.min(sub + 1, dp[am]);
        }
      }
    }
    return dp[dp.length - 1] == Integer.MAX_VALUE ? -1 : dp[dp.length - 1];
  }
}